File: leetcode/java/6.ZigZag-Conversion.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class Main {

    public static String convert(String s,int numRows){
        if(numRows==0||s.length()<numRows){
            return s;
        }
        int currentRow=0;
        boolean reverse=false;
        String result="";
        List<ArrayList<Character>> rows = new ArrayList<ArrayList<Character>>();
        for(int i=0;i<numRows;i++){
           rows.addAll(Collections.singleton(new ArrayList<Character>()));
        }
        
        for (int i=0;i<s.length();i++){
            rows.get(currentRow).add(s.charAt(i));
            if(!reverse){
                currentRow++;
            }else{
                currentRow--;
            }
            if(currentRow == numRows -1 || currentRow ==0 ){
                reverse = !reverse;
            }
        }

        for(int i=0;i<rows.size();i++){
        String str="";
        for(int j=0;j<rows.get(i).size();j++){
          str += rows.get(i).get(j);    
        }
        result+= str;
        }
        
        return result;
    }

    public static void main(String[] args) {
        System.out.println(convert("PAYPALISHIRING", 3));
    }
    
}

File: leetcode/kotlin/6.ZigZag-Conversion.kt
import java.util.*

fun convert(s:String, numRows:Int):String{

    if(numRows==0 || s.length <numRows){
        return s;
    }
    var currentRow:Int=0;
    var reverse:Boolean=false;
    var result:String="";

    var rows= mutableListOf<MutableList<Char>>()

    for(i  in 0..numRows){
        rows.addAll(Collections.singleton(mutableListOf()))
    }

    for (element in s){
        rows[currentRow].add(element)
        if(!reverse){
            currentRow++;
        }else{
            currentRow--;
        }
        if(currentRow == numRows -1 || currentRow ==0 ){
            reverse = !reverse;
        }
    }

    for (i in 0 until rows.size){
        var str=""
        for(j in 0 until rows[i].size){
        str += rows[i][j];
    }
        result+= str;
    }

    return result;
}

fun main(){
    println(convert("PAYPALISHIRING",3))
}